tls/s2n_cipher_suites.c

```diff
@@ -1155,24 +1155,179 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
     return 0;
 }
 
-static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+static int s2n_wire_ciphers_contain(const uint8_t *match, struct s2n_stuffer *wire_stuffer, uint32_t count, uint32_t cipher_suite_len)
 {
+    uint8_t theirs[S2N_SSLv2_CIPHER_SUITE_LEN] = { 0 };
+    POSIX_ENSURE_LTE(cipher_suite_len, sizeof(theirs));
+
     for (uint32_t i = 0; i < count; i++) {
-        const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
+        POSIX_GUARD(s2n_stuffer_read_bytes(wire_stuffer, theirs, cipher_suite_len));
 
         if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            POSIX_GUARD(s2n_stuffer_reread(wire_stuffer));
             return 1;
         }
     }
 
+    POSIX_GUARD(s2n_stuffer_reread(wire_stuffer));
     return 0;
 }
 
+static S2N_RESULT s2n_chacha20_is_in_cipher_preferences(const struct s2n_cipher_preferences *cipher_preferences, bool *chacha20_is_present) {
+    RESULT_ENSURE_REF(cipher_preferences);
+    RESULT_ENSURE_REF(chacha20_is_present);
+    *chacha20_is_present = false;
+
+    const uint8_t chacha20_iana[S2N_TLS_CIPHER_SUITE_LEN] = { TLS_CHACHA20_POLY1305_SHA256 };
+
+    for (int i = 0; i < cipher_preferences->count; i++) {
+        const uint8_t *cipher_iana = cipher_preferences->suites[i]->iana_value;
+        if (s2n_constant_time_equals(chacha20_iana, cipher_iana, S2N_TLS_CIPHER_SUITE_LEN)) {
+            *chacha20_is_present = true;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/* If the client has ChaCha20 as their most preferred cipher suite (first cipher suite) then set flag to true. */
+static S2N_RESULT s2n_wire_indicates_chacha20_boosting(struct s2n_stuffer *wire_stuffer, uint8_t count, uint32_t cipher_suite_len, bool *chacha20_boosted) {
```

Comment:
This seems pretty heavy-duty for what it's trying to do. Couldn't you just do `s2n_constant_time_equals(wire, s2n_tls13_chacha20_poly1305_sha256.iana_value, S2N_TLS_CIPHER_SUITE_LEN)`? Are stuffers and a separate method necessary here?

But does just comparing to s2n_tls13_chacha20_poly1305_sha256 / TLS_CHACHA20_POLY1305_SHA256 cover all use cases? What about the TLS1.2 chacha20 cipher suites?

Reply:
The idea was that in the future, we could extend this function to include different ciphers. But I re-factored the logic a bit, so this got pulled out and simplified.